includes/gateway/class-omise-payment-creditcard.php

```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
 		}
 
 		/**
-		 * @param  int $order_id
-		 *
-		 * @see    WC_Payment_Gateway::process_payment( $order_id )
-		 * @see    woocommerce/includes/abstracts/abstract-wc-payment-gateway.php
-		 *
-		 * @return array
+		 * @inheritdoc
 		 */
-		public function process_payment( $order_id ) {
-			if ( ! $order = $this->load_order( $order_id ) ) {
-				wc_add_notice(
-					sprintf(
-						wp_kses(
-							__( 'We cannot process your payment.<br/>Note that nothing wrong by you, this might be from our store issue.<br/><br/>Please feel free to try submit your order again or report our support team that you have found this problem (Your temporary order id is \'%s\')', 'omise' ),
-							array(
-								'br' => array()
-							)
-						),
-						$order_id
-					),
-					'error'
-				);
-				return;
+		public function charge( $order_id, $order ) {
+			$token   = isset( $_POST['omise_token'] ) ? wc_clean( $_POST['omise_token'] ) : '';
+			$card_id = isset( $_POST['card_id'] ) ? wc_clean( $_POST['card_id'] ) : '';
+
+			if ( empty( $token ) && empty( $card_id ) ) {
+				throw new Exception( __( 'Please select an existing card or enter a new card information.', 'omise' ) );
 			}
 
-			$order->add_order_note( __( 'Omise: Processing a payment with Credit Card solution..', 'omise' ) );
+			$user              = $order->get_user();
+			$omise_customer_id = $this->is_test() ? $user->test_omise_customer_id : $user->live_omise_customer_id;
 
-			try {
-				$token   = isset( $_POST['omise_token'] ) ? wc_clean( $_POST['omise_token'] ) : '';
-				$card_id = isset( $_POST['card_id'] ) ? wc_clean( $_POST['card_id'] ) : '';
-
-				if ( empty( $token ) && empty( $card_id ) ) {
-					throw new Exception( __( 'Please select an existing card or enter a new card information.', 'omise' ) );
+			// Saving card.
+			if ( isset( $_POST['omise_save_customer_card'] ) && empty( $card_id ) ) {
+				if ( empty( $token ) ) {
+					throw new Exception( __( 'Cannot process with a card that you\'re using. Please make sure that the card info is correct or contact our support team if you have any questions.', 'omise' ) );
```

Comment:
Unable to process the card. Please make sure that the information is correct, or contact our support team if you have any questions.